### PR TITLE
Bug fix using same seed for channel keys generation

### PR DESCRIPTION
**Pull request by yuntai**

### CODE REVIEW

**Comment by ariard** [comment]
Great, was what I would like to do at first, but didn't in keys interface rush writing for demo :)

**Comment by TheBlueMatt** [comment]
Oops, thanks for this, totally forgot to fix pre-merge. I think we should think about folding in RNG input as well here - the channel keys need backup anyway as they're mostly used in combination with remote input and we don't want an old backup to end up replaying keys.

**Comment by yuntai** [comment]
ah..  then, for channel keys, HD stuff is any helpful? hash(rng) is enough?

**Comment by TheBlueMatt** [comment]
If we have more stuff available, shoving it in the hash won't hurt and RNGs have been more than problematic on some platforms many times - hashing rng output plus something derived from the master key seems like a good idea. Maybe be extra doubly overkill careful and shove the current time in there as well to make sure keys can't be duplicated, even though it doesn't help with security.

**Comment by yuntai** [comment]
I'm still not so sure why we need to keep around a fixed channel master key (for a fixed user seed) if we randomize the derived keys anyway. To me, hash(rng||timestamp) as a seed for generating channel_keys alone seems enough and simple(no id tracking).

**Comment by TheBlueMatt** [comment]
One of the long-term goals of the key interface is to move towards a world where we can function without using our own RNG dependency and just push that complexity onto the user (for, eg, a hardware wallet or some other device with shitty entropy).

**Comment by yuntai** [comment]
@ariard Sorry, didn't realize you are in the middle of work for this. As this patch is a quick point fix, hope it won't conflict too much with your local work.

**Comment by ariard** [comment]
Ah sorry no I didn't work further on the channel key generation, because it's only a default implementation for KeysInterface. Long term it should be up to the user to write it's own scheme.
I'm writing tests for spendable outputs events, but nothing which should conflict with yours.

File: src/chain/keysinterface.rs [hunk 1]
```diff
@@ -192,11 +198,16 @@ impl KeysInterface for KeysManager {
...
-		let channel_pubkey = ExtendedPubKey::from_private(&self.secp_ctx, &self. channel_master_key);
-		let mut seed = [0; 32];
-		for (arr, slice) in seed.iter_mut().zip((&channel_pubkey.public_key.serialize()[0..32]).iter()) {
-			*arr = *slice;
-		}
-		ChannelKeys::new_from_seed(&seed)
+		let mut seed = [0u8; 32 + 4 + 8 + 33];
+		rng::fill_bytes(&mut seed[..32]);
+		let now = SystemTime::now().duration_since(UNIX_EPOCH).expect("Time went backwards");
+		seed[32..32+4].copy_from_slice(&byte_utils::be32_to_array(now.subsec_micros()));
```
Comment (TheBlueMatt) [inline src/chain/keysinterface.rs]: I think you want nanos.

File: src/chain/keysinterface.rs [hunk 2]
```diff
@@ -192,11 +198,16 @@ impl KeysInterface for KeysManager {
...
+		seed[32+4..32+4+8].copy_from_slice(&byte_utils::be64_to_array(now.as_secs()));
+
+		let child_ix = self.channel_child_index.fetch_add(1, Ordering::SeqCst);
```
Comment (TheBlueMatt) [inline src/chain/keysinterface.rs]: RelAcq should be fine here.

File: src/chain/keysinterface.rs [hunk 3]
```diff
@@ -192,11 +198,16 @@ impl KeysInterface for KeysManager {
...
+		let child_privkey = self.channel_master_key.ckd_priv(&self.secp_ctx, ChildNumber::from_hardened_idx(child_ix as u32)).expect("Your RNG is busted");
+		let child_pubkey = ExtendedPubKey::from_private(&self.secp_ctx, &child_privkey);
```
Comment (TheBlueMatt) [inline src/chain/keysinterface.rs]: Why convert to pub? Just hash the secret key?

File: src/chain/keysinterface.rs [hunk 4]
```diff
@@ -192,11 +198,16 @@ impl KeysInterface for KeysManager {
...
 	fn get_channel_keys(&self, _inbound: bool) -> ChannelKeys {
-		let channel_pubkey = ExtendedPubKey::from_private(&self.secp_ctx, &self. channel_master_key);
-		let mut seed = [0; 32];
-		for (arr, slice) in seed.iter_mut().zip((&channel_pubkey.public_key.serialize()[0..32]).iter()) {
-			*arr = *slice;
-		}
-		ChannelKeys::new_from_seed(&seed)
+		let mut seed = [0u8; 32 + 4 + 8 + 33];
+		rng::fill_bytes(&mut seed[..32]);
+		let now = SystemTime::now().duration_since(UNIX_EPOCH).expect("Time went backwards");
```
Comment (TheBlueMatt) [inline src/chain/keysinterface.rs]: Can you add a comment here something like:

"We only seriously intend to rely on the channel_master_key for true secure entropy, everything else just ensures uniqueness. We generally don't expect all clients to have non-broken RNGs here, so we also include the current time as a fallback to get uniqueness."

File: src/chain/keysinterface.rs [hunk 5]
```diff
@@ -192,11 +198,18 @@ impl KeysInterface for KeysManager {
...
+		let mut seed = [0u8; 32 + 4 + 8 + 32];
+		rng::fill_bytes(&mut seed[..32]);
+		let now = SystemTime::now().duration_since(UNIX_EPOCH).expect("Time went backwards");
+		seed[32..32+4].copy_from_slice(&byte_utils::be32_to_array(now.subsec_nanos()));
+		seed[32+4..32+4+8].copy_from_slice(&byte_utils::be64_to_array(now.as_secs()));
+
+		let child_ix = self.channel_child_index.fetch_add(1, Ordering::AcqRel);
+		let child_privkey = self.channel_master_key.ckd_priv(&self.secp_ctx, ChildNumber::from_hardened_idx(child_ix as u32)).expect("Your RNG is busted");
+		seed[32+4+8..32+4+8+32].copy_from_slice(&child_privkey.secret_key[..32]);
+		ChannelKeys::new_from_seed(&Sha256dHash::from_data(&seed).as_bytes())
```
Comment (TheBlueMatt) [inline src/chain/keysinterface.rs]: No need to double-sha here.

File: src/chain/keysinterface.rs [hunk 6]
```diff
@@ -192,11 +198,18 @@ impl KeysInterface for KeysManager {
...
+		// to have non-broken RNGs here, so we also include the current time as a fallback to get uniqueness."
+		let mut seed = [0u8; 32 + 4 + 8 + 32];
+		rng::fill_bytes(&mut seed[..32]);
+		let now = SystemTime::now().duration_since(UNIX_EPOCH).expect("Time went backwards");
+		seed[32..32+4].copy_from_slice(&byte_utils::be32_to_array(now.subsec_nanos()));
+		seed[32+4..32+4+8].copy_from_slice(&byte_utils::be64_to_array(now.as_secs()));
+
+		let child_ix = self.channel_child_index.fetch_add(1, Ordering::AcqRel);
+		let child_privkey = self.channel_master_key.ckd_priv(&self.secp_ctx, ChildNumber::from_hardened_idx(child_ix as u32)).expect("Your RNG is busted");
+		seed[32+4+8..32+4+8+32].copy_from_slice(&child_privkey.secret_key[..32]);
```
Comment (TheBlueMatt) [inline src/chain/keysinterface.rs]: nit: secret_key *is* 32 bytes.

File: src/chain/keysinterface.rs [hunk 7]
```diff
@@ -192,11 +198,18 @@ impl KeysInterface for KeysManager {
...
 	fn get_channel_keys(&self, _inbound: bool) -> ChannelKeys {
-		let channel_pubkey = ExtendedPubKey::from_private(&self.secp_ctx, &self. channel_master_key);
-		let mut seed = [0; 32];
-		for (arr, slice) in seed.iter_mut().zip((&channel_pubkey.public_key.serialize()[0..32]).iter()) {
-			*arr = *slice;
-		}
-		ChannelKeys::new_from_seed(&seed)
+		// We only seriously intend to rely on the channel_master_key for true secure entropy,
+		// everything else just ensures uniqueness. We generally don't expect all clients
+		// to have non-broken RNGs here, so we also include the current time as a fallback to get uniqueness."
```
Comment (TheBlueMatt) [inline src/chain/keysinterface.rs]: nit: Oops, you left the " at the end here.

**Pull request closed without merging**